### PR TITLE
Wide generic argument in function

### PR DESCRIPTION
**Pull request by valery1707**
For issue #1386

### CODE REVIEW

File: javaslang/src-gen/main/java/javaslang/CheckedFunction1.java [hunk 1]
```diff
@@ -79,7 +79,7 @@
      * @return a function that applies arguments to the given {@code partialFunction} and returns {@code Some(result)}
      *         if the function is defined for the given arguments, and {@code None} otherwise.
      */
-    static <T1, R> Function1<T1, Option<R>> lift(CheckedFunction1<T1, R> partialFunction) {
+    static <T1, R> Function1<T1, Option<R>> lift(CheckedFunction1<? super T1, R> partialFunction) {
```
Comment (danieldietrich) [inline javaslang/src-gen/main/java/javaslang/CheckedFunction1.java]: I think we will also need to widen the result type R: `? extends R` for al occurrences (lift, liftTry, recover).

Could you also widen the parameters (`? super`for args and `? extends` for return type) for the method `of(...)` of CheckedFunction and Function?

**Comment by valery1707** [comment]
@danieldietrich, fixed

**Comment by danieldietrich** [comment]
Thank you for applying my suggestions. Looking at the changes of `of`, `lift` and `liftTry` I'm double-minded if the bounds are really needed. I need to sleep one night about the topic. Maybe widening the bounds does not make sense for one or more of the methods I mentioned above.

Sorry for being undecided here.

**Comment by danieldietrich** [comment]
Okay, we are nearly finished.

Please revert the changes for the `of` Method. The function argument has to be the same type like the result function. The `of` method does not use generic type bounds.

Example:

```java
static <T1, T2, R> Function2<T1, T2, R> of(Function2<T1, T2, R> methodReference) {
    return methodReference;
}
```

Explanation:

The purpose of the `of` method is to turn a lambda expression into a Javaslang function. The arg is returned as-is. It does only work well with lambda expressions. The types of method references cannot be inferred by Java.

File: javaslang/src-gen/main/java/javaslang/CheckedFunction0.java [hunk 2]
```diff
@@ -75,8 +75,8 @@
      * @return a function that applies arguments to the given {@code partialFunction} and returns {@code Some(result)}
      *         if the function is defined for the given arguments, and {@code None} otherwise.
      */
-    static <R> Function0<Option<R>> lift(CheckedFunction0<R> partialFunction) {
-        return () -> Try.of(partialFunction::apply).getOption();
+    static <R> Function0<Option<R>> lift(CheckedFunction0<? extends R> partialFunction) {
+        return () -> Try.of(of(partialFunction)::apply).getOption();
```
Comment (danieldietrich) [inline javaslang/src-gen/main/java/javaslang/CheckedFunction0.java]: We can replace the method call `of(partialFunction)` by providing a type hint `<R>`:

```java
return () -> Try.<R> of(partialFunction::apply).getOption();
```

**Comment by valery1707** [comment]
@danieldietrich, fixed

**Comment by danieldietrich** [comment]
Great, thank you!

**Pull request merged**